lib/frontend/src/switch.rs

```diff
@@ -0,0 +1,193 @@
+use std::collections::HashMap;
+use cranelift_codegen::ir::*;
+use cranelift_codegen::ir::condcodes::IntCC;
+use frontend::FunctionBuilder;
+
+type EntryIndex = i64;
+
+#[derive(Debug)]
+pub struct Switch {
+    cases: HashMap<EntryIndex, Ebb>
+}
+
+impl Switch {
+    pub fn new() -> Self {
+        Switch {
+            cases: HashMap::new(),
+        }
+    }
+
+    pub fn set_entry(&mut self, index: EntryIndex, ebb: Ebb) {
+        assert!(self.cases.insert(index, ebb).is_none(), "Tried to set the same entry {} twice", index);
```

Comment:
Could you move the `insert` out of the `assert!` into a separate statement? The code works either way, but I find it easier to read when the program logic is separated from `assert!` conditions.